src/server/visual-diff-reporter.js

```diff
@@ -0,0 +1,111 @@
+import { cpSync, mkdirSync, rmSync, writeFileSync } from 'fs';
+import { dirname, join } from 'path';
+import { getTestInfo, PATHS } from './visual-diff-plugin.js';
+import { execSync } from 'child_process';
+import { fileURLToPath } from 'url';
+
+const __dirname = dirname(fileURLToPath(import.meta.url));
+
+function createData(rootDir, sessions) {
+
+	const files = new Map();
+	const browsers = new Map();
+
+	sessions.forEach(s => {
+
+		const browserName = s.browser.name;
+		if (!browsers.has(browserName)) {
+			browsers.set(browserName, {
+				name: browserName,
+				numFailed: 0,
+				version: s.browser.browser.version().substring(0, s.browser.browser.version().indexOf('.'))
+			});
+		}
+		const browserData = browsers.get(browserName);
+
+		const fileName = s.testFile.substring(rootDir.length + 1);
+		if (!files.has(fileName)) {
+			files.set(fileName, { name: fileName, numFailed: 0, tests: new Map() });
+		}
+		const fileData = files.get(fileName);
+		flattenResults(s, browserData, fileData);
+
+	});
+
+	return { files, browsers };
+
+}
+
+function flattenResults(session, browserData, fileData) {
+
+	function collectTests(prefix, tests) {
+		tests.forEach(t => {
+			const testName = `${prefix}${t.name}`;
+			const testKey = testName.replaceAll(' > ', ' ');
+			if (!fileData.tests.has(testName)) {
+				fileData.tests.set(testName, {
+					name: testName,
+					numFailed: 0,
+					results: []
+				});
+			}
+			const testData = fileData.tests.get(testName);
+			if (!t.passed) {
+				browserData.numFailed++;
+				fileData.numFailed++;
+				testData.numFailed++;
+			}
+			testData.results.push({
+				name: browserData.name,
+				duration: t.duration,
+				error: t.error?.message,
+				passed: t.passed,
+				info: getTestInfo(session, testKey)
+			});
+		});
+	}
+
+	function collectSuite(prefix, suite) {
+		collectTests(prefix, suite.tests);
+		suite.suites.forEach(s => collectSuite(`${prefix}${s.name} > `, s));
+	}
+
+	if (session.testResults) {
+		collectSuite('', session.testResults);
+	}
+
+}
+
+export function visualDiffReporter({ reportResults = true } = {}) {
+	let rootDir;
+	return {
+		start({ config }) {
+			rootDir = config.rootDir;
+		},
+		stop({ sessions }) {
```

Comment:
Does this work with `--watch`? I think I expected this in `onTestRunFinished` or `reportTestFileResults`, but looking at those again they seem to run for each test file which is no good 🤔. But I think ideally in `watch` mode you want a new report after each run

Reply:
Ok so I tried this out, and `stop` is not called at all in watch mode. Which kind of jives with their documentation for what it's supposed to do:
> Called once when the test runner stops. This can be used to write a test report to disk for regular test runs.

They call out "regular test runs" which I think implies "not watch runs". I think I'm OK with not writing a report when watching, since you can focus on particular files, go into manual mode and all kinds of other stuff.

Reply:
Yeah, when I was dev-ing in watch mode before reports were a thing I didn't really miss them because you're filtered down to just the tests you're working on and trying to quickly iterate and check out the `resized` or `diff` pngs.

But looking at the docs again, `onTestRunFinished` _might_ give us what we'd need, it does have `sessions`. Does it run only in watch mode and not in regular mode, or would putting the same functionality there end up running twice in regular mode?

Reply:
`onTestRunFinished` runs in both, and in watch mode it re-runs each time you change a file or focus on a particular file. I actually originally had all this in there, but moved it up into `stop` to avoid having to think about watch mode and reports with partial sessions. I think if it's something we want to support, it's definitely possible.

Reply:
Right, you'd only have the last session that ran passed to `onTestRunFinished`, but data for all the sessions that have been running and you'd probably want to create a full report still to be accurate. Yeah, I don't mind `--watch` clearing the report and not creating a new one, and worrying about supporting it later if we find we want it.